### DISCUSSION TITLE
Train commands

### DISCUSSION THREAD

**Discussion by carmel4a** [MaSzyna-Reloaded/MaSzyna-API-wrapper#41] [General]
Currently we are exposing _TrainPart_ behavior as _state_ Dictionary and _commands_. I don't see any advantage to expose _commands_ instead of raw C++ functions.

**Comment by marcinn** [comment]
Train commands exposed as strings is an abstraction layer for High-Level API, which is designed for communication between game objects and external entities. They're easy to pack in any transport layer. 

They're very similar to signals at c++ level. The only one difference is in GDSCript, where signals are exposed as first-class citizens (under the hood they're still string names).  

You can still communicate directly through methods instead of sending commands  (as native C++ methods, but also via Godot/GDExtension/GDScript interface). This style is provided for internal communication, where internal structure of a high-level component is well known.
 
Please check W.I.P. docs here: 
* https://marcinn.github.io/MaSzyna-API-wrapper/architecture.html
* https://marcinn.github.io/MaSzyna-API-wrapper/wrapping-mover.html#command-callbacks

### Problematic example

Note that  `TrainPart` can provide own behavior, incl. command handlers. But TrainPart node can be nested in a quite free form in the vehicle composition, including it's unique node name. Imagine two different vehicles, and assume that both support `engine_start` command:

```
+ SM42 (TrainController)
    +---  StonkaEngine
```

and

```
+ SM42_V2
    +--- NewDieselElectricEngine
```

To start the engine, In first case you will call `$SM42/StonkaEngine.engine_start()`. For the second case it will be: `$SM42_V2/NewDieselElectricEngine.engine_start()`. Now try to implement input handler for `E` keydown action, which will always call `engine_start()`. Without knowing the internal structure of the vehicle, you will not be able to call `engine_start()`.

Alternatively, all commands must be implemented in `TrainController` (at High-Level Component level). But because of the nature of the commands (they're dynamic), and because we're relying on a composition of multiple TrainParts, all these methods would have to be defined at runtime. In such case, all calls of these methods will need to be wrapped with `has_method` checks, and you will have no possibility to call c++ native methods. 

### Constants

The current design is a good compromise for flexibility. To avoid mistakes and make the code less error prone, a constants will be introduced. I.e. instead of:
```gdscript
TrainSystem.send_command("sm42", "engine_start")
```
you will able to write:
```gdscript
TrainSystem.send_command("sm42", TrainEngine.COMMAND_ENGINE_START)
```
or
```gdscript
TrainSystem.send_command("sm42", TrainEngine.Commands.ENGINE_START)
```

### Extensibility

Imagine a some Vehicle, which has no support for `cabin_ligths` command. You're still add it to the scene, add new `CustomPartNode` and register `cabin_lights` callback. The method will be defined, for example, in `Vehicle/CustomPartNode.toggle_cabin_lights()`, but it will be accessible through unified interface: 
```gdscript
TrainSystem.send_command("vehicle", "cabin_lights", true)
```
The same call, as for any other vehicle. So if you already hav an input handler for `L` key, which sends `cabin_lights` command, the `toggle_cabin_lights()` method will be called for your customized vehicle.

Overriding commands is not supported yet, but we can consider such possibility via `TrainSystem.replace_command()`.

### Broadcasting and rerouting

Sending commands through dispatcher like `TrainSystem` gives a possibility to broadcasting commands, re-routing, re-sending or queuing them. It will be quite easy to implement radio commands, which will spread as a very simplified radio waves. It is already possible to broadcast a command to all vehicles. A dispatcher will be also able to re-send the command to all cars in a trainset.

The biggest benefit comes from decoupling the dispatch logic from the command handlers. You can easily implement RadioEmitter for simulating radio commands (i.e. radio stop or radio messages). 

### Summing up

So there is a quite difference between a command callback and sending a command.:

* TrainSystem allows to send commands identified by string names. Commands can be delivered or not, can be broadcasted, queued, or send again.
* Pure C++ methods allows to execute a some logic immediately. This can be expected in some cases, of course.

> _NOTE_: without CabinSystem, game events and communication between game objects it's hard to say yet which layer should be exposed as commands. We currently expose Mover methods, but in the future commands may be more related to the Cabin interface (for example).
